### PR TITLE
Allow marking sender to never show images

### LINKED ISSUE #3146: Missing load external image buttons.
In mobile layout the load external image buttons are not always visible or they are not recognizable by the user. The notifications and buttons are also shown differently in mobile / desktop layout. We should find a common way to display the information. 

1. Show images is not available when the mail is not authenticated. A click/touch on the text will show the load images button button but it is also not clear that this is a clickable item.
authentication-fail-mobile

2. People don't recognize that "Show images" is actual a clickable button:
 always-trust-mobile

I would prefer to show all messages at once and add buttons like in the desktop layout or at least use a regular button style.
load-images-desktop

### PR DESCRIPTION
**Pull request by johnbotris**
This depends on #3175

closes #3146

### CODE REVIEW

**Review by vitoreiji — CHANGES_REQUESTED** [review]
I noticed this doesn't implement "an option in settings where a user can see trusted and blocked external senders" as proposed in the issue. Has this been decided against? If not, then maybe we could just release it like this and postpone implementing the list. Otherwise, maybe we could consider hashing instead of encrypting.

File: src/mail/view/MailViewer.js [hunk 1]
```diff
@@ -880,9 +892,9 @@ export class MailViewer {
 			throw e
 		}
 
-		const isAllowListedExternalSender = await worker.isAllowedExternalSender(mail.sender.address)
-		                                                .catch(e => false)
-		const isAllowedAndAuthenticatedExternalSender = isAllowListedExternalSender
+		const externalImageRule = await worker.getExternalImageRule(mail.sender.address)
+		                                      .catch(e => ExternalImageRule.None)
```
[the remark this answers is not in the document]
Comment (johnbotris) [reply src/mail/view/MailViewer.js:896]: it should only be dberrors in this case, but it's just for safety

File: src/mail/view/MailViewer.js [hunk 2]
```diff
@@ -1641,25 +1660,44 @@ export class MailViewer {
...
-					locator.search.indexingSupported && this.mail.authStatus === MailAuthenticationStatus.AUTHENTICATED
-						? {
-							text: "allowExternalContentSender_action",
-							click: () => this._setContentBlockingStatus(ContentBlockingStatus.AlwaysShow)
-						}
-						: null
-				]
-			})
-			: null
+		if (this._contentBlockingStatus !== ContentBlockingStatus.Block) {
```
[the remark this answers is not in the document]
Comment (johnbotris) [reply src/mail/view/MailViewer.js:1663]: it should be!

File: src/mail/view/MailViewer.js [hunk 3]
```diff
@@ -1641,25 +1660,44 @@ export class MailViewer {
...
+				{
+					label: "blockExternalContentSender_action",
+					click: () => this._setContentBlockingStatus(ContentBlockingStatus.AlwaysBlock)
+				}
+			].filter(Boolean)
+			: []
+
+		// on narrow screens the buttons will end up on 2 lines if there are too many, this looks bad.
+		const maybeDropdownButtons = styles.isSingleColumnLayout() && alwaysOrNeverAllowButtons.length > 1
+			? [ moreButton(alwaysOrNeverAllowButtons, 216) ]
```
[the remark this answers is not in the document]
Comment (johnbotris) [reply src/mail/view/MailViewer.js:1689]: width of the dropdown

File: src/mail/view/MailViewer.js [hunk 4]
```diff
@@ -1671,18 +1709,18 @@ export class MailViewer {
...
-		} else if (previousContentBlockingStatus === ContentBlockingStatus.AlwaysShow) {
-			// if we're going from allow to something else it means we're revoking the whitelisting of the given sender
-			worker.removeAllowedExternalSender(this.mail.sender.address).catch(IndexingNotSupportedError, noOp)
+		if (status === ContentBlockingStatus.AlwaysShow) {
+			worker.addExternalImageRule(this.mail.sender.address, ExternalImageRule.Allow).catch(IndexingNotSupportedError, noOp)
+		} else if (status === ContentBlockingStatus.AlwaysBlock) {
+			worker.addExternalImageRule(this.mail.sender.address, ExternalImageRule.Block).catch(IndexingNotSupportedError, noOp)
+		} else {
+			// we are going from allow or block to something else it means we're resetting to the default rule for the given sender
+			worker.addExternalImageRule(this.mail.sender.address, ExternalImageRule.None).catch(IndexingNotSupportedError, noOp)
```
[the remark this answers is not in the document]
Comment (johnbotris) [reply src/mail/view/MailViewer.js:1718]: not available yet!

**Comment by johnbotris** [comment]
> I noticed this doesn't implement "an option in settings where a user can see trusted and blocked external senders" as proposed in the issue. Has this been decided against? If not, then maybe we could just release it like this and postpone implementing the list. Otherwise, maybe we could consider hashing instead of encrypting.

IMO it would be a good thing to have in the future

**Comment by charlag** [comment]
What's the status? Needs to be reviewed again?

**Review by charlag — APPROVED** [review]

**Pull request merged**